design/src/rfcs/rfc0012_body_callback_apis.md

```diff
@@ -0,0 +1,297 @@
+RFC: Callback APIs for `ByteStream` and `SdkBody`
+=================================================
+
+> Status: RFC
+
+Adding a callback API to `ByteStream` and `SdkBody` will enable developers using the SDK to implement things like checksum validations and 'read progress' callbacks.
+
+## The Implementation
+
+*Note that comments starting with '//' are not necessarily going to be included in the actual implementation and are intended as clarifying comments for the purposes of this RFC.*
+
+```rust
+// in aws_smithy_http::callbacks...
+
+/// A callback that, when inserted into a request body, will be called for corresponding lifecycle events.
+// Docs for these methods will mostly be the same as the docs on `Callback` so I've omitted them.
+trait BodyCallback: Send {
+   /// This lifecycle function is called for each chunk **successfully** read. If an error occurs while reading a chunk,
+   /// this will not be called. This function takes `&mut self` so that implementors may modify an implementing
+   /// struct/enum's internal state. Implementors may return an error.
+   fn update(&mut self, #[allow(unused_variables)] bytes: &[u8]) -> Result<(), BoxError> { Ok(()) }
+
+   /// This callback is called once all chunks have been read. If the callback encountered 1 or more errors
+   /// while running `update`s, this is how those errors are raised. Implementors may return a [`HeaderMap`][HeaderMap]
+   /// that will be appended to the HTTP body as a trailer. This is only useful to do for streaming requests.
+   fn trailers(&self) -> Result<Option<HeaderMap<HeaderValue>>, BoxError> { Ok(None) }
```

Comment:
I think that for `SdkBody` I can have it inspect itself and only append trailers if the inner variant is `Inner::Streaming` but I'm not sure that covers every streaming request case.